selectFields is being ignored

I haven't got selectFields ever working, hash and salt is being included on each query.

```javascript
var mongoose = require('mongoose');
var Schema = mongoose.Schema;
var util = require('util');
var assert = require('assert');
var passportLocalMongoose = require('passport-local-mongoose');

mongoose.set('debug', true);
mongoose.connect('mongodb://localhost/gh2735');

var UserSchema = new mongoose.Schema({
  username: { type: String, lowercase: true, required: true, select: true },
  email: { type: String, lowercase: true, required: true, select: true },
  salt: { type: String, select: false },
  hash: { type: String, select: false }
});


UserSchema.plugin(passportLocalMongoose, {
  selectFields: 'username email -hash -salt'
});

var UserModel = mongoose.model('User', UserSchema);

var user = new UserModel({ username: 'vkarpov15', email: 'a@b.co', salt: 'yes please', hash: 'LV426' });

user.save(function(error, user) {
    assert.ifError(error);
    UserModel.findOne({ _id: user._id })
    .exec(function(error, user) {
      assert.ifError(error);
      assert.ok(!user.salt);
      assert.ok(!user.hash);
      console.log(util.inspect(user));
      process.exit(0);
    });
  });
```

this code outputs:

    Mongoose: users.insert({ username: 'vkarpov15', email: 'a@b.co', salt: 'yes please', hash: 'LV426', _id: ObjectId("551bd2af69fda1530e95d63c"), __v: 0 })   
    Mongoose: users.findOne({ _id: ObjectId("551bd2af69fda1530e95d63c") }) { fields: undefined }  

    assert.js:86
      throw new assert.AssertionError({
            ^
    AssertionError: false == true
        at EventEmitter.<anonymous> (/home/jussi/development/mongoosebug/gh-2735.js:32:14)
        at EventEmitter.<anonymous> (/home/jussi/development/mongoosebug/node_modules/mongoose/node_modules/mpromise/lib/promise.js:175:45)
        at EventEmitter.emit (events.js:107:17)
        at Promise.safeEmit (/home/jussi/development/mongoosebug/node_modules/mongoose/node_modules/mpromise/lib/promise.js:81:21)
        at Promise.fulfill (/home/jussi/development/mongoosebug/node_modules/mongoose/node_modules/mpromise/lib/promise.js:94:24)
        at Promise.resolve (/home/jussi/development/mongoosebug/node_modules/mongoose/lib/promise.js:113:23)
        at /home/jussi/development/mongoosebug/node_modules/mongoose/lib/query.js:1169:16
        at /home/jussi/development/mongoosebug/node_modules/mongoose/node_modules/kareem/index.js:103:16
        at process._tickCallback (node.js:355:11)

Example code taken from: https://github.com/Automattic/mongoose/issues/2735

--- a/lib/passport-local-mongoose.js
+++ b/lib/passport-local-mongoose.js
@@ -53,8 +53,8 @@ module.exports = function(schema, options) {
     if (!schema.path(options.usernameField)) {
         schemaFields[options.usernameField] = { type : String, unique : options.usernameUnique };
     }
-    schemaFields[options.hashField] = String;
-    schemaFields[options.saltField] = String;
+    schemaFields[options.hashField] = {type: String, select: false};
+    schemaFields[options.saltField] = {type: String, select: false};
 
     if (options.limitAttempts){
       schemaFields[options.attemptsField] = {type: Number, default: 0};
@@ -166,7 +166,7 @@ module.exports = function(schema, options) {
         var self = this;
 
         return function(username, password, cb) {
-            self.findByUsername(username, function(err, user) {
+            self.findByUsername(username, true, function(err, user) {
                 if (err) { return cb(err); }
 
                 if (user) {
@@ -226,8 +226,12 @@ module.exports = function(schema, options) {
         });
     };
 
-    schema.statics.findByUsername = function(username, cb) {
+    schema.statics.findByUsername = function(username, selectHashSaltFields, cb) {
         var queryParameters = {};
+        if (typeof cb === 'undefined') {
+          cb = selectHashSaltFields;
+          selectHashSaltFields = false;
+        }
 
         // if specified, convert the username to lowercase
         if (username !== undefined && options.usernameLowerCase) {
@@ -237,6 +241,11 @@ module.exports = function(schema, options) {
         queryParameters[options.usernameField] = username;
 
         var query = this.findOne(queryParameters);
+
+        if (selectHashSaltFields) {
+            query.select('+' + options.hashField + " +" + options.saltField);
+        }
+
         if (options.selectFields) {
             query.select(options.selectFields);
         }

--- a/test/issues.js
+++ b/test/issues.js
@@ -79,12 +79,10 @@ describe('issues', function () {
 
                 var authenticate = User.authenticate();
 
-                authenticate('hugo', 'password', function(err, auth, reason) {
+                authenticate('hugo', 'password', function(err, result) {
                     assert.ifError(err);
-                    assert.equal(false, auth);
-                    assert.ok(reason);
 
-                    assert.equal('Authentication not possible. No salt value stored in mongodb collection!', reason.message);
+                    assert.ok(result instanceof User);
 
                     done();
                 });
@@ -147,6 +145,28 @@ describe('issues', function () {
         User.register({username: "nicolascage"}, 'password', function (err, user) {
             assert.equal(err, "My nasty error");
             done();
+        });
+    });
+
+    it('should not expose hash and salt fields - Issue #72', function(done) {
+        this.timeout(5000); // Five seconds - mongo db access needed
+
+        var UserSchema = new Schema({});
+
+        UserSchema.plugin(passportLocalMongoose, { });
+        var User = mongoose.model('ShouldNotExposeHashAndSaltFields_Issue_72', UserSchema);
+
+        User.register({username: "nicolascage"}, 'password', function (err, user) {
+            assert.ifError(err);
+            assert.ok(user);
+            User.findOne({username: "nicolascage"}, function(err, user) {
+                assert.ifError(err);
+                assert.ok(user);
+                assert.equal(user.username, "nicolascage");
+                assert.strictEqual(user.hash, undefined);
+                assert.strictEqual(user.salt, undefined);
+                done();
+            });
         });
     });
 });

